File: objective-c/HYLImageManager/HYLFileManager/HYLFileManager.m
//
//  HYLFileManager.m
//The MIT License (MIT)
//
//Copyright (c) 2015 Yilei He
//https://github.com/lionhylra/HYLImageManager
//
//Permission is hereby granted, free of charge, to any person obtaining a copy
//of this software and associated documentation files (the "Software"), to deal
//in the Software without restriction, including without limitation the rights
//to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//copies of the Software, and to permit persons to whom the Software is
//furnished to do so, subject to the following conditions:
//
//The above copyright notice and this permission notice shall be included in all
//copies or substantial portions of the Software.
//
//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//SOFTWARE.

#import "HYLFileManager.h"
NSString *const kDefaultPathComponent = @"UserDocuments";

@interface HYLFileManager()

@end

@implementation HYLFileManager

-(instancetype)initWithDirectory:(NSSearchPathDirectory)directory rootPathComponents:(nullable NSArray *)pathComponents {
    self = [super init];
    if (self) {
        _rootPathComponents = pathComponents;
        _directory = directory;
    }
    return self;
}

-(instancetype)init{
    return [self initWithDirectory:kDefaultDirectory rootPathComponents:@[kDefaultPathComponent]];
}

+(instancetype)defaultManager{
    static dispatch_once_t pred;
    static id sharedInstance = nil;
    dispatch_once(&pred, ^{
        sharedInstance = [[[self class] alloc] init];
    });
    return sharedInstance;
}

-(NSString *)pathForFileName:(NSString *)fileName{
    if (fileName == nil) {
        return nil;
    }
    if ([fileName isKindOfClass:[NSNull class]]) {
        return nil;
    }
    if (fileName.length == 0) {
        return nil;
    }
    NSString *filePath = [self buildPathInSystemDirectory:NSSearchPathForDirectoriesInDomains(self.directory, NSUserDomainMask, YES).firstObject];
    filePath = [filePath stringByAppendingPathComponent:fileName];
    return filePath;
}

-(NSData *)loadDataWithName:(NSString * __nonnull)fileName{
    return [NSData dataWithContentsOfFile:[self pathForFileName:fileName]];
}

-(void)saveData:(NSData * __nonnull)data withName:(NSString * __nonnull)fileName{
    NSString *path = [self pathForFileName:fileName];
    [self validateDirectoryPath:[path stringByDeletingLastPathComponent]];
    [data writeToFile:path atomically:YES];
}

-(BOOL)deleteFileWithName:(NSString * __nonnull)fileName error:(NSError **)error{
    NSString *filePath = [self pathForFileName:fileName];
    if(![[NSFileManager defaultManager] fileExistsAtPath:filePath]){
        return YES;
    }
    
    return [[NSFileManager defaultManager] removeItemAtPath:filePath error:error];
}

-(BOOL)renameFileFromFileName:(NSString * __nonnull)oldName toNewFileName:(NSString * __nonnull)newName error:(NSError *__autoreleasing  __nullable * __nullable)error{
    NSString *oldPath = [self pathForFileName:oldName];
    NSString *newPath = [self pathForFileName:newName];
    [self validateDirectoryPath:[newPath stringByDeletingLastPathComponent]];
    return [[NSFileManager defaultManager] moveItemAtPath:oldPath toPath:newPath error:error];
}

-(BOOL)fileExistsForFileName:(NSString * __nonnull)fileName{
    return [[NSFileManager defaultManager] fileExistsAtPath:[self pathForFileName:fileName]];
}
#pragma mark - private mathods

-(NSString *)buildPathInSystemDirectory:(NSString *)directoryPath{
    NSString *filePath = directoryPath;
    if (self.rootPathComponents) {
        for (NSString *component in self.rootPathComponents) {
            filePath = [filePath stringByAppendingPathComponent:component];
        }
    }
    
    return filePath;
}

-(void)validateDirectoryPath:(NSString *)path{
    if (!path) {
        return;
    }
    BOOL isDirectory;
    if([[NSFileManager defaultManager] fileExistsAtPath:path isDirectory:&isDirectory]&&isDirectory){
        return;
    }
    /* create directory */
    NSError *error;
    BOOL isSuccess = [[NSFileManager defaultManager] createDirectoryAtPath:path withIntermediateDirectories:YES attributes:nil error:&error];
    if (!isSuccess) {
        NSLog(@"Create directory error: %@", error.localizedDescription);
    }
    
}



@end

File: swift/HYLImageManager/HYLFileManager/HYLFileManager.swift
//
//  HYLFileManager.swift
//  HYLMultiImagesViewControllerDemo
//
//  Created by HeYilei on 21/08/2015.
//  Copyright (c) 2015 HeYilei. All rights reserved.
//

import UIKit

public let kDefaultPathComponent = "UserDocuments"
public let kDefaultDirectory = NSSearchPathDirectory.DocumentDirectory

public class HYLFileManager: NSObject {
    // MARK: - Public Properties
    public let rootPathComponents:[String]?
    public let directory:NSSearchPathDirectory
    public static let defaultManager = HYLFileManager()
    
    // MARK: - Initializers
    init(directory:NSSearchPathDirectory, rootPathComponents:[String]?){
        self.rootPathComponents = rootPathComponents
        self.directory = directory
        super.init()
    }
    
    convenience override init() {
        self.init(directory: kDefaultDirectory, rootPathComponents: [kDefaultPathComponent])
    }
    
    // MARK: - File Operation Methods
    public func pathForFileName(fileName:String) -> String?{
        if fileName.isEmpty {
            return nil
        }
        let filePath = buildPathInSystemDirectory(NSSearchPathForDirectoriesInDomains(self.directory, NSSearchPathDomainMask.UserDomainMask, true)[0] )
        return filePath.stringByAppendingPathComponent(fileName)
    }
    
    public func loadDataWithName(fileName:String) -> NSData? {
        if let filePath = pathForFileName(fileName) {
            return NSData(contentsOfFile: filePath)
        }
        return nil
    }
    
    public func saveData(data:NSData, withName fileName:String){
        if let path = pathForFileName(fileName) {
            validateDirectoryPath(path.stringByDeletingLastPathComponent)
            data.writeToFile(path, atomically: true)
        }
    }
    
    public func deleteFileWithName(fileName:String) throws{
        let error: NSError! = NSError(domain: "Migrator", code: 0, userInfo: nil)
        if let path = pathForFileName(fileName) {
            if !NSFileManager.defaultManager().fileExistsAtPath(path) {
                return
            }
            
            try NSFileManager.defaultManager().removeItemAtPath(path)
            return
        }
        throw error
    }
    
    public func renameFileFrom(oldName:String, toNewFileName newName:String) throws{
        let error: NSError! = NSError(domain: "Migrator", code: 0, userInfo: nil)
        if let oldPath = pathForFileName(oldName), newPath = pathForFileName(newName) {
            validateDirectoryPath(newPath.stringByDeletingLastPathComponent)
            try NSFileManager.defaultManager().moveItemAtPath(oldPath, toPath: newPath)
            return
        }
        throw error
    }
    
    public func fileExistsForFileName(fileName:String) -> Bool {
        if let path = pathForFileName(fileName){
            return NSFileManager.defaultManager().fileExistsAtPath(path)
        }
        return false
    }
    
    // MARK: - Private Methods
    private func buildPathInSystemDirectory(directoryPath:String) -> String{
        var filePath = directoryPath
        if let pathComponents = self.rootPathComponents {
            for component in pathComponents{
                filePath = filePath.stringByAppendingPathComponent(component)
            }
        }
        return filePath
    }
    
    private func validateDirectoryPath(path:String){
        var isDirectory = ObjCBool(false)
        if NSFileManager.defaultManager().fileExistsAtPath(path, isDirectory: &isDirectory) && isDirectory {
            return
        }
        
        var error:NSError?
        do {
            try NSFileManager.defaultManager().createDirectoryAtPath(path, withIntermediateDirectories: true, attributes: nil)
        } catch let error1 as NSError {
            error = error1
            print("Create directory for HYLFileManager failed with error: \(error!.localizedDescription)")
        }
    }
}
